### DISCUSSION TITLE
Track dialing progress

### DISCUSSION THREAD

**Discussion by drHuangMHT** [libp2p/rust-libp2p#4987] [Q&A] [answered]
According to the coding guideline:
https://github.com/libp2p/rust-libp2p/blob/b7914e407da34c99fb76dcc300b3d44b9af97fac/docs/coding-guidelines.md?plain=1#L304-L313
Shouldn't there be a way to track the progress of a dialing request issued to the swarm? How can you track that request when `swarm.dial()` returns `Ok(())` when the request is accepted?

**Comment by thomaseizinger** [comment]
The `DialOpts` you pass in give you access to the `ConnectionId`. You can then correlate all events to this call to `Swarm::dial`. Because the same API is used in `ToSwarm::dial`, we opted to not return `ConnectionId` because it is information the caller already knows :)

**Reply by drHuangMHT** [reply]
Why this design when there's no such thing as `ListenOpts`? I think it's an inconsistency when `listen_on` gives you the `ListenerId` upon accepted. Besides, `ConnectionId` is more of an internal state, so it should be managed, or in this case assigned, internally. It feels like we are managing it externally when we need to get it from `DialOpts`. Moreover, this can potentially create unused `ConnectionId` when a `DialOpts` is built but not used. It is an case that basically never happens and does no harm to the program, but it can lead to confusion with logs in which I would naturally assume the `ConnectionId` increments continously.
In a word, it makes *more sense* when `Swarm::dial()` return `Ok(ConnectionId)` rather than obtaining it from `DialOpts`.

**Reply by thomaseizinger** [reply, marked as answer]
We didn't want to return information from a function that was passed in and we had to create the `ConnectionId` within `DialOpts` because it was a requirement to obtain the `ConnectionId` from a yet-to-be-created connection within a `NetworkBehaviour` to ensure `NetworkBehaviour`s can learn, which connections they initiated via `ToSwarm::Dial`.

**Reply by drHuangMHT** [reply]
> which connections they initiated via `ToSwarm::Dial`.

I see, so it's a problem of the architecture.
